account_global_discount/models/account_move.py

```diff
@@ -179,27 +181,25 @@ def _recompute_global_discount_lines(self):
                 }
             )
 
-    @api.onchange("partner_id", "company_id")
-    def _onchange_partner_id(self):
-        res = super()._onchange_partner_id()
-        discounts = False
-        if (
-            self.move_type in ["out_invoice", "out_refund"]
-            and self.partner_id.customer_global_discount_ids
-        ):
-            discounts = self.partner_id.customer_global_discount_ids.filtered(
-                lambda d: d.company_id == self.company_id
-            )
-        elif (
-            self.move_type in ["in_refund", "in_invoice"]
-            and self.partner_id.supplier_global_discount_ids
-        ):
-            discounts = self.partner_id.supplier_global_discount_ids.filtered(
-                lambda d: d.company_id == self.company_id
-            )
-        if discounts:
-            self.global_discount_ids = discounts
-        return res
+    @api.depends("partner_id", "company_id", "move_type")
+    def _compute_global_discount_ids(self):
+        for move in self:
+            discounts = self.env["global.discount"]
+            move_discounts = self.env["global.discount"]
+            if (
+                move.move_type in ["out_invoice", "out_refund"]
+                and move.partner_id.customer_global_discount_ids
+            ):
+                move_discounts = move.partner_id.customer_global_discount_ids
+            elif (
+                move.move_type in ["in_refund", "in_invoice"]
+                and move.partner_id.supplier_global_discount_ids
+            ):
+                move_discounts = move.partner_id.supplier_global_discount_ids
+            for discount in move_discounts:
+                if discount.company_id == move.company_id:
+                    discounts |= discount
```

Comment:
For company filtering `filtered` expressions, as old code does, could be used instead another `for` statement here

Reply:
In the internal process of the "filtered" function, a "browse" is performed. In my opinion, this approach is more efficient than using the "filtered" function itself.